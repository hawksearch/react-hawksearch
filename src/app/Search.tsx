//
// this file is the entry point for the main search application when running the minified bundle
//

import '@babel/polyfill';
import 'url-search-params-polyfill';

import React from 'react';
import ReactDOM from 'react-dom';

import 'rheostat/initialize';

import QueryStringListener from 'components/QueryStringListener';
import App from 'app/App';
import Hawksearch from 'components/Hawksearch';
import { parseLocation } from 'util/QueryString';
import { HawksearchConfig } from 'types/HawksearchConfig';

declare global {
	interface Window {
		Hawksearch?: HawksearchConfig;
	}
}

if (window.Hawksearch) {
	initializeSearch(window.Hawksearch);
}

export function initializeSearch(config: HawksearchConfig) {
	if (!config.searchElement) {
		throw new Error('window.Hawksearch.searchElement has not been configured');
	}

	const renderElem =
		typeof config.searchElement === 'string' ? document.getElementById(config.searchElement) : config.searchElement;

	// pull the initial search from location
	const searchRequest = parseLocation(location);

	ReactDOM.render(
		<Hawksearch config={config} initialSearch={searchRequest}>
			<QueryStringListener />
			<App />
		</Hawksearch>,
		renderElem
	);
}
