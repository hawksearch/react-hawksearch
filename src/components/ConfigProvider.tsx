import React, { useContext } from 'react';

import { HawksearchConfig } from 'types/HawksearchConfig';

const ConfigContext = React.createContext({} as ConfigContextValue);

export interface ConfigContextValue {
	/** Global configuration. */
	config: HawksearchConfig;
}

export interface ConfigProviderProps {
	config: HawksearchConfig;
	children: React.ReactNode;
}

// declare webpack's constant to appease typescript
declare var __webpack_public_path__: string;

function ConfigProvider({ config, children }: ConfigProviderProps) {
	if (config.assetPath) {
		let path = config.assetPath;

		// ensure the provided path both starts and ends with a slash
		if (!path.startsWith('/')) {
			path = '/' + path;
		}
		if (!path.endsWith('/')) {
			path = path + '/';
		}

		// allow consumers to tell webpack where to load code split/lazy loaded files from, as they may not be
		// hosting our JS from /assets/ (the default path)
		__webpack_public_path__ = path;
	}

	return <ConfigContext.Provider value={{ config }}>{children}</ConfigContext.Provider>;
}

export function useHawkConfig() {
	const context = useContext(ConfigContext);

	if (!context.config) {
		throw new Error(
			'No HawksearchConfig is available, did you forget to wrap your components in a ConfigProvider component?'
		);
	}

	return context;
}

export default ConfigProvider;
