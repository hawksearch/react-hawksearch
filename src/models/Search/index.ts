export * from './Pagination';
export * from './Result';
export * from './Request';
export * from './Response';
export * from './Selections';
export * from './SmartBar';
export * from './Sorting';
export * from './ClientData';
export * from './Merchandising';
export * from './PageContent';
export * from './ContentType';
export * from './Trigger';
export * from './Rule';
