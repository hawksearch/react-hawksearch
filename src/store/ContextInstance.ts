import React from 'react';

export function createContext(id: string, defaultValue: any) {
    if (!window.HawksearchReact) {
        window.HawksearchReact = {};
    }

    if (!window.HawksearchReact.contextInstances) {
        window.HawksearchReact.contextInstances = {};
    }

    if (window.HawksearchReact.contextInstances[id]) {
        console.log('reused context ' + id)
        return window.HawksearchReact.contextInstances[id];
    }
    else {
        let context = React.createContext(defaultValue);
        console.log('created new context')
        window.HawksearchReact.contextInstances[id] = context;

        return context;
    }
}