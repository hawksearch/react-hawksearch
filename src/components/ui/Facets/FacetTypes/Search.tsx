import React, { useState } from 'react';

import { useHawkSearch } from 'components/StoreProvider';
import { useFacet } from 'components/ui/Facets/Facet';
import { useTranslation } from 'react-i18next';

function Search() {
	const { store, actor: hawkActor } = useHawkSearch();
	const { facet, actor } = useFacet();

	const [input, setInput] = useState<string | undefined>(undefined);

	const { t, i18n } = useTranslation();

	function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
		if (event.key === 'Enter') {
			setInput(undefined); // clear the user's entered value as we want to be driven by the store again

			actor.selectFacet(event.currentTarget.value);
		}
	}

	function clearFacet() {
		setInput(undefined); // clear the user's entered value as we want to be driven by the store again

		hawkActor.clearFacet(facet);
	}

	function getInputValue() {
		if (input !== undefined) {
			// if the user type an input, that's the value for the input
			return input;
		}

		// otherwise, use the value from the store
		return store.pendingSearch.SearchWithin || '';
	}

	return (
		<>
			<div className="hawk-facet-rail__facet-values">
				<div className="hawk-facet-rail__facet-values__search">
					<input
						value={getInputValue()}
						onChange={e => setInput(e.currentTarget.value)}
						onKeyDown={onKeyDown}
					/>
				</div>
			</div>
			{store.pendingSearch.SearchWithin && (
				<div className="hawk-facet-rail__facet-values__search-clear">
					<button className="link-button" onClick={clearFacet}>
						{t('Clear')}
					</button>
				</div>
			)}
		</>
	);
}

export default Search;
