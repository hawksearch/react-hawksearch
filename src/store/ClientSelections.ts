import { Facet } from 'models/Facets';

export interface ClientSelections {
	[field: string]: ClientSelection;
}

export interface ClientSelection {
	/** The facet that this selection is for. */
	facet: Facet;

	/** The display name for the facet. */
	label: string;

	items: ClientSelectionValue[];
}

export interface ClientSelectionValue {
	/** Display label for facet value. */
	label: string;

	/** Value for facet value. */
	value: string;
	path?: string;
}
