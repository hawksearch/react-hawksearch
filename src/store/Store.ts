import { Response, Request, Result } from 'models/Search';
import { Value, Facet } from 'models/Facets';
import { ClientSelections, ClientSelectionValue } from 'store/ClientSelections';

export enum FacetSelectionState {
	/** The facet value is not selected. */
	NotSelected,
	/** The facet value is selected. */
	Selected,
	/** The facet value is selected, but negated. */
	Negated,
}

export interface SelectionInfo {
	/** The facet value selection state. */
	state: FacetSelectionState;

	/**
	 * If the facet value is `FacetSelectionState.Selected` or `FacetSelectionState.Negated`, this is the value of
	 * the facet value. For negated facet values this will be prefixed with the negation character `'-'`.
	 */
	selectedValue?: string;
	/**
	 * If the facet value is `FacetSelectionState.Selected` or `FacetSelectionState.Negated`, this is the index
	 * into the `pendingSearch.FacetSelections[facetName]` array for this facet value.
	 */
	selectionIndex?: number;
}

export class SearchStore {
	/** This represents the next search request that will be executed. */
	public pendingSearch: Partial<Request>;
	/**
	 * Whether or not the next search request will perform history actions (pushing the search into browser
	 * history).
	 */
	public doHistory: boolean;

	/** Whether or not a search request is waiting for completion. */
	public isLoading: boolean;

	public itemsToCompare: Result[];
	public itemsToCompareIds: string[];
	public comparedResults: Result[];

	/** The results of the last search request, if one has been performed. Otherwise, `undefined`. */
	public searchResults?: Response;

	public requestError: boolean;

	public constructor(initial?: Partial<SearchStore>) {
		Object.assign(this, initial);
	}

	/**
	 * Returns whether or not this is the initial load of the search results.
	 */
	public get isInitialLoad() {
		return this.isLoading && !this.searchResults;
	}

	/**
	 * Determines whether or not the given facet and facet value is selected, and returns info regarding the selection.
	 * @param facet The facet for which the facet value will be checked for selection.
	 * @param facetValue The facet value that will be checked for selection.
	 */
	public isFacetSelected(facet: Facet | string, facetValue: Value | string): SelectionInfo {
		const facetName = typeof facet === 'string' ? facet : facet.Name;
		const facetField = typeof facet === 'string' ? facet : facet.selectionField;

		const valueValue = typeof facetValue === 'string' ? facetValue : facetValue.Value;
		const valueLabel = typeof facetValue === 'string' ? facetValue : facetValue.Label;

		if (!valueValue) {
			console.error(`Facet ${facetName} (${facetField}) has no facet value for ${valueLabel}`);
			return { state: FacetSelectionState.NotSelected };
		}

		const facetSelections = this.pendingSearch.FacetSelections;

		if (!facetSelections || !facetSelections[facetField]) {
			return { state: FacetSelectionState.NotSelected };
		}

		const selectionIdx = facetSelections[facetField]!.indexOf(valueValue);
		const negationIdx = facetSelections[facetField]!.indexOf(`-${valueValue}`);

		if (selectionIdx !== -1) {
			// if the exact facet value exists, then we're normally selected
			return {
				state: FacetSelectionState.Selected,
				selectedValue: valueValue,
				selectionIndex: selectionIdx,
			};
		} else if (negationIdx !== -1) {
			// if the facet value is selected but prefixed with a -, then we're negated
			return {
				state: FacetSelectionState.Negated,
				selectedValue: `-${valueValue}`,
				selectionIndex: negationIdx,
			};
		}

		return { state: FacetSelectionState.NotSelected };
	}

	/**
	 * Returns an object containing the selections that have been made in both the next search request and also
	 * in the previous search request. This should be used when iterating selections instead of pulling the values
	 * out from the search result or pending search - as this will merge the values together and provide an accurate
	 * view of all facet selections.
	 */
	public get facetSelections(): ClientSelections {
		const {
			pendingSearch: { FacetSelections: clientSelections, SearchWithin },
			searchResults,
		} = this;
		const selections: ClientSelections = {};

		if (!clientSelections && !SearchWithin) {
			return selections;
		}

		// if we've made selections on the client, transform these into more detailed selections.
		// the client-side selections are just facet fields and values without any labels - so we
		// need to combine this information with the list of facets received from the server in the
		// previous search in order to return a rich list of selections

		const facets = searchResults ? searchResults.Facets : null;

		if (!facets) {
			// but we can only do this if we've received facet information from the server. without this
			// info we can't determine what labels should be used
			return selections;
		}

		// manually handle the `searchWithin` selection, as this doesn't usually behave like a normal facet selection
		// but instead a field on the search request
		if (SearchWithin) {
			const facet = facets.find(f => f.selectionField === 'searchWithin');

			if (facet) {
				selections.searchWithin = {
					facet,
					label: facet.Name,
					items: [
						{
							label: SearchWithin,
							value: SearchWithin,
						},
					],
				};
			}
		}

		if (!clientSelections) {
			return selections;
		}

		Object.keys(clientSelections).forEach(fieldName => {
			const selectionValues = clientSelections[fieldName];

			if (!selectionValues) {
				// if this selection has no values, it's not really selected
				return;
			}

			const facet = facets.find(f => f.selectionField === fieldName);

			if (!facet) {
				// if there's no matching facet from the server, we can't show this since we'll have no labels
				return;
			}

			const items: ClientSelectionValue[] = [];

			if (facet.FieldType === 'range') {
				// if the facet this selection is for is a range, there won't be a matching value and thus there won't be a label.
				// so because of this we'll just use the selection value as the label

				selectionValues.forEach(selectionValue => {
					items.push({
						label: selectionValue,
						value: selectionValue,
					});
				});
			} else if (facet.FieldType === 'tab') {
				// do not return the selection value for tab facet
				return;
			} else {
				// for other types of facets, try to find a matching value

				selectionValues.forEach(selectionValue => {
					const matchingVal = this.findMatchingValue(selectionValue, facet.Values);

					if (!matchingVal || !matchingVal.Label) {
						// if there's no matching value from the server, we cannot display because there would
						// be no label - same if there's no label at all
						return;
					}

					items.push({
						label: matchingVal.Label,
						value: selectionValue,
						path: matchingVal.Path,
					});
				});
			}

			selections[fieldName] = {
				facet,
				label: facet.Name,
				items,
			};
		});

		return selections;
	}

	private findMatchingValue(selectionValue: string, facetValues: Value[]): Value | null {
		let matchingValue: Value | null = null;
		if (!facetValues || facetValues.length === 0) {
			return null;
		}

		for (const facetValue of facetValues) {
			const isMatchingVal = facetValue.Value === selectionValue || `-${facetValue.Value}` === selectionValue;
			// loop through children
			if (!isMatchingVal) {
				matchingValue = this.findMatchingValue(selectionValue, facetValue.Children);
			} else {
				matchingValue = facetValue;
			}

			if (matchingValue) {
				return matchingValue;
			}
		}

		return matchingValue;
	}
}
