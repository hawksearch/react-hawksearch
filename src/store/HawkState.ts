import { useEffect } from 'react';
import axios, { CancelToken } from 'axios';

import { SearchStore, FacetSelectionState } from './Store';
import HawkClient from 'net/HawkClient';
import { Response, Request, FacetSelections } from 'models/Search';
import { useMergableState } from 'util/MergableState';
import { useHawkConfig } from 'components/ConfigProvider';
import { Facet, Value } from 'models/Facets';
import { FacetType } from 'models/Facets/FacetType';

export interface SearchActor {
	/**
	 * Performs a search with the currently configured pending search request. The search request can be
	 * configured via the `setSearch` method. This method usually doesn't need to be called directly, as
	 * the `StoreProvider` component will usually trigger searches directly in response to calls to
	 * `setSearch`.
	 * @returns A promise that resolves when the search request has been completed.
	 */
	search(cancellationToken?: CancelToken): Promise<void>;

	/**
	 * Configures the next search request that will be executed. This will also execute a search in response to
	 * the next search request changing.
	 * @param search The partial search request object. This will be merged with previous calls to `setSearch`.
	 * @param doHistory Whether or not this search request will push a history entry into the browser. If
	 * 					not specified, the default is `true`.
	 */
	setSearch(search: Partial<Request>, doHistory?: boolean): void;

	/**
	 * Toggles a facet value for the next search request that will be executed. If the given facet had previously
	 * been selected, it will be unselected. If the negation state of a selected facet is changed, the selection
	 * will have its negation state changed. Internally, this will call `setSearch` to configure the next search
	 * with this selected facet.
	 * @param facet The facet for which the value is being selected.
	 * @param facetValue The facet value being selected.
	 * @param negate  Whether or not this selection is considered a negation.
	 */
	toggleFacetValue(facet: Facet | string, facetValue: Value | string, negate?: boolean): void;

	setFacetValues(facet: Facet | string, facetValues: Value[] | string[]): void;

	/**
	 * Entirely clears all the values of the given facet from the current selection.
	 * @param facet The facet to clear.
	 */
	clearFacet(facet: Facet | string): void;

	/**
	 * Clears a given facet value of the given facet from the current selection.
	 * @param facet The facet to clear.
	 * @param facetValue The facet value to clear.
	 */
	clearFacetValue(facet: Facet | string, facetValue?: Value | string): void;

	/**
	 * Clears all selected facets from the current selection.
	 */
	clearAllFacets(): void;
}

export function useHawkState(initialSearch?: Partial<Request>): [SearchStore, SearchActor] {
	const { config } = useHawkConfig();

	const client = new HawkClient(config);

	const [store, setStore] = useMergableState(
		new SearchStore({
			pendingSearch: initialSearch || {
				FacetSelections: {},
			},
			isLoading: true,
		}),
		SearchStore
	);

	useEffect(() => {
		// when the pending search changes, trigger a search

		const cts = axios.CancelToken.source();
		search(cts.token);

		return () => {
			cts.cancel();
		};
	}, [store.pendingSearch]);

	/**
	 * Performs a search with the currently configured pending search request. The search request can be
	 * configured via the `setSearch` method. This method usually doesn't need to be called directly, as
	 * the `StoreProvider` component will usually trigger searches directly in response to calls to
	 * `setSearch`.
	 * @returns A promise that resolves when the search request has been completed.
	 */
	async function search(cancellationToken?: CancelToken): Promise<void> {
		setStore({ isLoading: true });
		let searchResults: Response | null = null;

		try {
			searchResults = await client.search(
				{
					// the search request being executed is spread from the pendingSearch
					...store.pendingSearch,
					// pass parameter for extended response
					IsInPreview: config.isInPreview,
					// and override some of the request fields with config values
					ClientGuid: config.clientGuid,
				},
				cancellationToken
			);
		} catch (error) {
			if (axios.isCancel(error)) {
				// if the request was cancelled, it's because this component was updated
				return;
			}

			console.error('Search request error:', error);
		}

		setStore({ isLoading: false });

		if (searchResults) {
			if (!searchResults.Success) {
				console.error('Search result error:', searchResults);
			} else {
				setStore({
					searchResults: new Response(searchResults),
				});
			}
		}
	}

	/**
	 * Configures the next search request that will be executed. This will also execute a search in response to
	 * the next search request changing.
	 * @param search The partial search request object. This will be merged with previous calls to `setSearch`.
	 * @param doHistory Whether or not this search request will push a history entry into the browser. If
	 * 					not specified, the default is `true`.
	 */
	function setSearch(pendingSearch: Partial<Request>, doHistory?: boolean): void {
		if (doHistory === undefined) {
			doHistory = true;
		}

		setStore(prevState => {
			const newState = {
				pendingSearch: { ...prevState.pendingSearch, ...pendingSearch },
				doHistory,
			};
			if (newState.pendingSearch.Keyword === '') {
				newState.pendingSearch.Keyword = undefined;
			}

			return newState;
		});
	}

	/**
	 * Sets the facet selections and search within configuration for the next search request. This will also
	 * clear the page number of the next request to display the first page of results.
	 * @param selections The facet selections to set.
	 * @param searchWithin The search within value to set.
	 */
	function setSearchSelections(selections?: FacetSelections, searchWithin?: string) {
		setSearch({
			FacetSelections: selections,
			SearchWithin: searchWithin,

			// when we change facet selections, also clear the current page so that we navigate
			// back to the first page of results
			PageNo: undefined,
		});
	}

	/**
	 * Toggles a facet value for the next search request that will be executed. Internally, this will call
	 * `setSearch` to configure the next search with this selected facet.
	 * @param facet The facet for which the value is being selected.
	 * @param facetValue The facet value being selected.
	 * @param negate  Whether or not this selection is considered a negation.
	 */
	function toggleFacetValue(facet: Facet | string, facetValue: Value | string, negate?: boolean): void {
		if (negate === undefined) {
			negate = false;
		}

		const facetName: string = typeof facet === 'string' ? facet : facet.Name;
		const facetField = typeof facet === 'string' ? facet : facet.selectionField;

		const valueValue = typeof facetValue === 'string' ? facetValue : facetValue.Value;
		const valueLabel = typeof facetValue === 'string' ? facetValue : facetValue.Label;

		if (!valueValue) {
			console.error(`Facet ${facetName} (${facetField}) has no facet value for ${valueLabel}`);
			return;
		}

		let facetSelections = store.pendingSearch.FacetSelections;

		// handle `searchWithin` facet, which isn't a facet selection but is instead a field on the
		// search request.
		if (facetField === 'searchWithin') {
			// set the search within to the facet value provided
			setSearchSelections(facetSelections, /* searchWithin */ valueValue);

			return;
		}

		if (!facetSelections) {
			facetSelections = {};
		}

		if (!facetSelections[facetField]) {
			facetSelections[facetField] = [];
		}

		const { state: selState, selectionIndex } = store.isFacetSelected(facet, facetValue);

		if (selState === FacetSelectionState.Selected || selState === FacetSelectionState.Negated) {
			// we're selecting this facet, and it's already selected

			// first, remove it from our selections
			facetSelections[facetField]!.splice(selectionIndex!, 1);

			if (
				(selState === FacetSelectionState.Selected && negate) ||
				(selState === FacetSelectionState.Negated && !negate)
			) {
				// if we're toggling from negation to non-negation or vice versa, then push the new selection
				facetSelections[facetField]!.push(negate ? `-${valueValue}` : valueValue);
			} else {
				// if we're not toggling the negation, nothing to do because we already removed the selection above
			}
		} else {
			// not selected, so we want to select it
			facetSelections[facetField]!.push(negate ? `-${valueValue}` : valueValue);
		}

		if (facetSelections[facetField]!.length === 0) {
			// clean up any facets that no longer have any selected facet values
			delete facetSelections[facetField];
		}

		setSearchSelections(facetSelections, store.pendingSearch.SearchWithin);
	}

	function setFacetValues(facet: Facet | string, facetValues: Value[] | string[]): void {
		const facetName = typeof facet === 'string' ? facet : facet.Name;
		const facetField = typeof facet === 'string' ? facet : facet.selectionField;

		let facetSelections = store.pendingSearch.FacetSelections;

		if (!facetSelections) {
			facetSelections = {};
		}

		facetSelections[facetField] = [];

		for (const facetValue of facetValues) {
			const valueValue = typeof facetValue === 'string' ? facetValue : facetValue.Value;
			const valueLabel = typeof facetValue === 'string' ? facetValue : facetValue.Label;

			if (!valueValue) {
				console.error(`Facet ${facetName} (${facetField}) has no facet value for ${valueLabel}`);
				return;
			}

			facetSelections[facetField]!.push(valueValue);
		}

		setSearchSelections(facetSelections, store.pendingSearch.SearchWithin);
	}

	/**
	 * Entirely clears all the values of the given facet from the current selection.
	 * @param facet The facet to clear.
	 */
	function clearFacet(facet: Facet | string) {
		const facetField = typeof facet === 'string' ? facet : facet.selectionField;

		const facetSelections = store.pendingSearch.FacetSelections;

		// handle `searchWithin` facet, which isn't a facet selection but is instead a field on the
		// search request.
		if (facetField === 'searchWithin') {
			// set searchWithin to undefined to clear it
			setSearchSelections(facetSelections, /* searchWithin */ undefined);

			return;
		}

		if (!facetSelections || !facetSelections[facetField]) {
			// if there are no facet selections or the facet isn't selected at all, there's nothing to clear
			return;
		}

		delete facetSelections[facetField];

		setSearchSelections(facetSelections, store.pendingSearch.SearchWithin);
	}

	/**
	 * Clears a given facet value of the given facet from the current selection.
	 * @param facet The facet to clear.
	 * @param facetValue The facet value to clear.
	 */
	function clearFacetValue(facet: Facet | string, facetValue: Value | string) {
		const facetName = typeof facet === 'string' ? facet : facet.Name;
		const facetField = typeof facet === 'string' ? facet : facet.selectionField;

		const valueValue = typeof facetValue === 'string' ? facetValue : facetValue.Value;
		const valueLabel = typeof facetValue === 'string' ? facetValue : facetValue.Label;

		if (!valueValue) {
			console.error(`Facet ${facetName} (${facetField}) has no facet value for ${valueLabel}`);
			return;
		}

		// handle `searchWithin` facet, which isn't a facet selection but is instead a field on the
		// search request.
		if (facetField === 'searchWithin') {
			// set searchWithin to undefined to clear it
			setSearchSelections(store.pendingSearch.FacetSelections, /* searchWithin */ undefined);

			return;
		}

		const { state: selState, selectionIndex } = store.isFacetSelected(facet, facetValue);

		if (selState === FacetSelectionState.NotSelected) {
			// if there are no facet selections or the facet isn't selected at all, there's nothing to clear
			return;
		}

		const facetSelections = store.pendingSearch.FacetSelections!;

		// remove it from the selections
		facetSelections[facetField]!.splice(selectionIndex!, 1);

		if (facetSelections[facetField]!.length === 0) {
			// clean up any facets that no longer have any selected facet values
			delete facetSelections[facetField];
		}

		setSearchSelections(facetSelections, store.pendingSearch.SearchWithin);
	}

	/**
	 * Clears all selected facets from the current selection.
	 */
	function clearAllFacets(): void {
		setSearchSelections(undefined, undefined);
	}

	const actor: SearchActor = {
		search,
		setSearch,
		toggleFacetValue,
		setFacetValues,
		clearFacet,
		clearFacetValue,
		clearAllFacets,
	};

	return [store, actor];
}
