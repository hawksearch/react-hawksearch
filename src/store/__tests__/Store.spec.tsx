import React, { useEffect } from 'react';
import { CancelToken } from 'axios';
import TestRenderer from 'react-test-renderer';

import { useHawkState } from 'store/HawkState';
import HawkClient from 'net/HawkClient';
import { Request, Response } from 'models/Search';
import ConfigProvider from 'components/ConfigProvider';

jest.mock('net/HawkClient');
const HawkClientMock = HawkClient as jest.Mock<HawkClient>;

const searchMock = jest.fn(
	(request: Request, cancellationToken?: CancelToken): Promise<Response> => {
		const response = {
			Success: true,
			Results: [
				{
					DocId: '123',
					Document: { a: ['b'] },
					Score: 1.0,
				},
			],
			Facets: [
				{
					FacetId: 123,
					Name: '123',
					Field: '123',
					FieldType: 'search',
					FacetType: 'checkbox',
					DisplayType: 'default',
					MaxCount: 10,
					MinHitCount: 1,
					SortBy: 'name',
					ExpandSelection: false,
					IsNumeric: false,
					IsCurrency: false,
					IsCollapsible: true,
					IsCollapsedDefault: false,
					IsVisible: true,
					IsSearch: false,
					ScrollHeight: 100,
					ScrollThreshold: 200,
					TruncateThreshold: 200,
					SearchThreshold: 200,
					AlwaysVisible: true,
					SortOrder: 1,
					NofVisible: 10,
					SwatchData: [],
					FacetRangeDisplayType: 100,
					PreloadChildren: true,
					ShowSliderInputs: false,
					Ranges: [],
					Values: [],
				},
			],
			Pagination: {
				CurrentPage: 0,
				Items: [],
				MaxPerPage: 24,
				NofPages: 1,
				NofResults: 1,
			},
			Sorting: {
				Items: [],
			},
			DidYouMean: ['test'],
			Keyword: request.Keyword || '',
			SearchDuration: 100,
			Selections: {},
		};

		return Promise.resolve<Response>(new Response((response as unknown) as Response));
	}
);

HawkClientMock.mockImplementation(() => {
	return ({
		baseUrl: '',
		search: searchMock,
		autocomplete: jest.fn(),
	} as unknown) as HawkClient;
});

describe('Store', () => {
	it('performs a search', async done => {
		let searchPromise;

		// arrange
		function StoreComponent() {
			const [store, actor] = useHawkState();

			// act
			useEffect(() => {
				actor.setSearch({
					Keyword: 'this is my keyword',
				});

				searchPromise = actor.search();
			}, []);

			return <div>{JSON.stringify(store, null, 2)}</div>;
		}

		function TestedComponent() {
			return (
				<ConfigProvider config={{ clientGuid: '123' }}>
					<StoreComponent />
				</ConfigProvider>
			);
		}

		const renderer = TestRenderer.create(<TestedComponent />);

		// trigger an update so that the `useEffect` is triggered
		renderer.update(<TestedComponent />);

		await searchPromise;

		// assert
		expect(searchMock).toHaveBeenCalledWith(
			expect.objectContaining({ Keyword: 'this is my keyword' }),
			expect.anything()
		);
		expect(renderer.toJSON()).toMatchSnapshot();

		done();
	});
});
