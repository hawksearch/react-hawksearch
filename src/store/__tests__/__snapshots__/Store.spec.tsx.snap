// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store performs a search 1`] = `
<div>
  {
  "pendingSearch": {
    "FacetSelections": {},
    "Keyword": "this is my keyword"
  },
  "doHistory": true,
  "isLoading": false,
  "itemsToCompare": [],
  "itemsToCompareIds": [],
  "comparedResults": [],
  "searchResults": {
    "Success": true,
    "Pagination": {
      "NofResults": 1,
      "CurrentPage": 0,
      "MaxPerPage": 24,
      "NofPages": 1,
      "Items": []
    },
    "Keyword": "this is my keyword",
    "Results": [
      {
        "DocId": "123",
        "Score": 1,
        "Document": {
          "a": [
            "b"
          ]
        }
      }
    ],
    "Facets": [
      {
        "FacetId": 123,
        "Name": "123",
        "Field": "123",
        "FieldType": "search",
        "FacetType": "checkbox",
        "DisplayType": "default",
        "MaxCount": 10,
        "MinHitCount": 1,
        "SortBy": "name",
        "ExpandSelection": false,
        "IsNumeric": false,
        "IsCurrency": false,
        "IsCollapsible": true,
        "IsCollapsedDefault": false,
        "IsVisible": true,
        "IsSearch": false,
        "ScrollHeight": 100,
        "ScrollThreshold": 200,
        "TruncateThreshold": 200,
        "SearchThreshold": 200,
        "AlwaysVisible": true,
        "SortOrder": 1,
        "NofVisible": 10,
        "SwatchData": [],
        "FacetRangeDisplayType": 100,
        "PreloadChildren": true,
        "ShowSliderInputs": false,
        "Ranges": [],
        "Values": []
      }
    ],
    "Selections": {},
    "Sorting": {
      "Items": []
    },
    "DidYouMean": [
      "test"
    ],
    "Merchandising": {},
    "FeaturedItems": {},
    "SearchDuration": 100,
    "PageContent": []
  },
  "requestError": false
}
</div>
`;
